mne_qt_browser/tests/test_pg_specific.py

```diff
@@ -145,30 +147,20 @@ def test_ch_specific_annot_display(raw_orig, pg_backend):
     assert annot_dock.start_bx.value() == 4
     assert single_channel_annot.lower.xData[0] == 4
 
-    fig.close()
-
-
-@pytest.mark.skipif(
-    not hasattr(mne.viz._figure.BrowserBase, "_toggle_single_channel_annotation"),
-    reason="needs MNE 1.8",
-)
-def test_ch_specific_annot_interactions(raw_orig, pg_backend):
-    """Test interactiveness and responsiveness of channel specific annotations."""
-    ch_names = ["MEG 0133", "MEG 0142", "MEG 0143", "MEG 0423"]
-    annot_onset, annot_dur = 1, 2
-    annots = Annotations([annot_onset], [annot_dur], "some_chs", ch_names=[ch_names])
-    raw_orig.set_annotations(annots)
-
-    fig = raw_orig.plot()
-    fig.test_mode = True
-    annot = fig.mne.regions[0]
-
-    fig._fake_keypress("a")  # activate annotation mode
+    # MNE >= 1.8
+    if not check_version("mne", "1.8"):
+        warning_message = (
+            "must update MNE to >= 1.8 to test single channel annots interactions"
+        )
+        # emit a warning if the user tries to test single channel annots
+        with pytest.warns(UserWarning, match=warning_message):
+            warnings.warn(warning_message, UserWarning)
+            return
```

Comment:
Here you are testing that the warning you are emitting in the test function is emitted, which is not very interesting :smile: The idea is that if you run this test with MNE < 1.8 and run a `_fake_click` function that would add a channel specific annotation, it will issue the warning you added in `_pg_figure` and it won't raise an exception:

```python
        if not hasattr(self.weakmain(), "_toggle_single_channel_annotation"):
            warn(
                "MNE must be updated to version 1.8 or above "
                "to support add/remove channels from annotation. "
            )
```

i.e. add the same `_fake_click` command which adds a channel specific annotation within the `pytest.warns` context manager.

Reply:
Sorry I'm not clearly understanding. Are you thinking something like this?

```
    if not hasattr(mne.viz._figure.BrowserBase, "_toggle_single_channel_annotation"):
        warn(
            "MNE must be updated to version 1.8 or above "
            "to support add/remove channels from annotation. "
        )
    else:    
        # test if shift click an existing annotation removes object
        ch_index = np.mean(annot.single_channel_annots["MEG 0133"].ypos).astype(int)
        fig._fake_click(
            (4 + 2 / 2, ch_index),
            xform="data",
            button=1,
            modifier=Qt.ShiftModifier,
        )
        assert "MEG 0133" not in annot.single_channel_annots.keys()
    
        # test if shift click on channel adds annotation
        fig._fake_click(
            (4 + 2 / 2, ch_index),
            xform="data",
            button=1,
            modifier=Qt.ShiftModifier,
        )
        assert "MEG 0133" in annot.single_channel_annots.keys()

    fig.close()
    return
```

Should there be a context manager with pytest anywhere or does this suffice? I thought @larsoner suggested something of this form:

```
    if not check_version("mne", "1.8"):
        warning_message = (
            "must update MNE to >= 1.8 to test single channel annots interactions"
        )
        # emit a warning if the user tries to test single channel annots
        with pytest.warns(UserWarning, match=warning_message):
            warnings.warn(warning_message, UserWarning)
            return
```

Reply:
No, in `_pg_figure`, i.e. in the code of the browser, you should emit a warning if the version of MNE is not compatible with a feature available in the version of browser. Imagine if a users runs a browser version compatible with channel-specific annotation but with MNE 1.6. The code within `_pg_figure` will attempt to run the channel-specific annotation part, call `self.weakmain()._toggle_single_channel_annotation` and crash. This is why in the code of `_pg_figure` you need to check for the existence of the attribute `_toggle_single_channel_annotation` in `self.weakmain()` and issue a warning if it does not exist. This is what you did in 0bb8196ec9e3f7288774b4237868e0c0456a9165 in the modification to `_pg_figure`.

Now, in the tests, consider the following test function which checks that a `_fake_click` creates a channel specific annotations:

```python
def test_channel_specific_annotation():
    """Test that channel specific annotation works as intended."""
    raw = ...  # our raw objects that will be modified by the browser
    fig = ...  # the opened browser
    _fake_click(...)  # fake click which creates a channel specific annotation.
    # verify with an assertion that the channel specific annotations was created.
    assert "xxx" in raw.annotations.channel_specific
``` 

We have 2 scenarios: either we run this test function witth MNE 1.8, in which case it will work and pass; or we run it with an old MNE version and it will raise an error and fail. Now, since you modified `_pg_figure` with a warning emited for old MNE version, it will not raise an error anymore but issue the warning (and still fail as we automatically turn warnings into errors in our tests). Thus, you need to differentiate those 2 scenarios in the tests and to verify that the warning emited by `_pg_figure` is indeed present.

```python
def test_channel_specific_annotation():
    """Test that channel specific annotation works as intended."""
    raw = ...  # our raw objects that will be modified by the browser
    fig = ...  # the opened browser
    if check_version("mne", "1.8"):
        _fake_click(...)  # fake click which creates a channel specific annotation.
        # verify with an assertion that the channel specific annotation was created.
        assert "xxx" in raw.annotations.channel_specific
    else:  # old MNE
        with pytest.warns(RuntimeWarning, match="MNE must be updated to version 1.8 or above"):
            _fake_click(...)  # fake click which creates a channel specific annotation, same as before.
        # verify with an assertion that the channel specific annotation was NOT created
        assert "xxx" not in raw.annotations.channel_specific  
``` 

Let us know if it's still unclear, you are getting there! 
FYI, I might not be able to join today, I'm at a workshop abroad all day; and I will be on vacation for 2 weeks as of next Thursday.